src/main/kotlin/no/nav/familie/ef/sak/beregning/BeregningBarnetilsynService.kt

```diff
@@ -0,0 +1,112 @@
+
+        val barnetilsynMåneder = utgiftsperioder.map {
+            it.split()
+        }.flatMap {
+            it.map { utgiftsMåned ->
+                utgiftsMåned.tilBeløpsperiodeBarnetilsynDto(kontantstøttePerioder, tilleggsstønadsperioder)
+            }
+        }
+        return barnetilsynMåneder.mergeSammenhengendePerioder()
+    }
+}
+
+/**
+ * Del opp utgiftsperioder i atomiske deler (mnd).
+ * Eksempel: 1stk UtgiftsperiodeDto fra januar til mars deles opp i 3:
+ * listOf(UtgiftsMåned(jan), UtgiftsMåned(feb), UtgiftsMåned(mars))
+ */
+fun UtgiftsperiodeDto.split(): List<UtgiftsMåned> {
+    val perioder = mutableListOf<UtgiftsMåned>()
+    var måned = this.årMånedFra
+    while (måned.isBefore(this.årMånedTil) || måned.equals(this.årMånedTil)) {
+        perioder.add(UtgiftsMåned(måned, this.barn, this.utgifter))
+        måned = måned.plusMonths(1)
+    }
+    return perioder
+}
+
+/**
+ * Merger sammenhengende perioder hvor beløp og @BeløpsperiodeBarnetilsynDto#beregningsgrunnlag (it.toKey()) er like.
+ */
+fun List<BeløpsperiodeBarnetilsynDto>.mergeSammenhengendePerioder(): List<BeløpsperiodeBarnetilsynDto> {
+    return mapNotNull { it }.groupingBy { it.toKey() }
+            .aggregate { _, akkumulatorListe: MutableList<BeløpsperiodeBarnetilsynDto>?, nestePeriodeDto, first ->
+                when (first) {
+                    true -> mutableListOf(nestePeriodeDto)
+                    false -> oppdaterEllerLeggTilNy(akkumulatorListe, nestePeriodeDto)
+                }
+            }.values.flatten()
+}
+
+private fun oppdaterEllerLeggTilNy(akkumulatorListe: MutableList<BeløpsperiodeBarnetilsynDto>?,
+                                   nestePeriodeDto: BeløpsperiodeBarnetilsynDto) =
+        when (erSammenhengende(akkumulatorListe!!.last().periode, nestePeriodeDto.periode)) {
+            true -> lagNyTildato(akkumulatorListe, nestePeriodeDto)
+            false -> akkumulatorListe.leggTil(nestePeriodeDto)
+        }
+
+private fun lagNyTildato(akkumulatorListe: MutableList<BeløpsperiodeBarnetilsynDto>,
+                         nestePeriodeDto: BeløpsperiodeBarnetilsynDto): MutableList<BeløpsperiodeBarnetilsynDto> {
+    val oppdatertBeløpsperiodeKopi = lagKopiMedNyTildato(akkumulatorListe.last(), nestePeriodeDto)
+    return akkumulatorListe.byttUtSisteMed(oppdatertBeløpsperiodeKopi)
+}
+
+private fun lagKopiMedNyTildato(beløpsperiodeBarnetilsynDto: BeløpsperiodeBarnetilsynDto,
+                                nestePeriodeDto: BeløpsperiodeBarnetilsynDto): BeløpsperiodeBarnetilsynDto {
+    val nyPeriode = beløpsperiodeBarnetilsynDto.periode.copy(tildato = nestePeriodeDto.periode.tildato)
+    return beløpsperiodeBarnetilsynDto.copy(periode = nyPeriode)
+}
+
+private fun erSammenhengende(gjeldendePeriode: Periode,
+                             nestePeriode: Periode) =
+        gjeldendePeriode.tildato.month.equals(nestePeriode.fradato.minusMonths(1).month)
```

Comment:
Denne delen synes jeg var litt komplisert, er det mulig å gjøre noe slik i stedet? 
Hvis det nå er det samme, då får man fjernet `leggTil` og `byttUtSisteMed` også

```suggestion
fun List<BeløpsperiodeBarnetilsynDto>.mergeSammenhengendePerioder(): List<BeløpsperiodeBarnetilsynDto> {
    return this.fold(mutableListOf()) { acc, entry ->
        val last = acc.lastOrNull()
        if (last != null && erSammenhengendeMedSammeBeløp(last, entry)) {
            acc.removeLast()
            acc.add(last.copy(periode = last.periode.copy(tildato = entry.periode.tildato)))
        } else {
            acc.add(entry)
        }
        acc
    }
}

fun erSammenhengendeMedSammeBeløp(first: BeløpsperiodeBarnetilsynDto, second: BeløpsperiodeBarnetilsynDto) =
        first.periode.tildato.plusMonths(1) == second.periode.fradato &&
        first.toKey() == second.toKey()
```

Reply:
Liker fold() ovenfor.  
Vi hadde en versjon med fold i starten. Husker ikke helt hvorfor vi forkastet denne, men det er mulig vi har refaktorert annen kode np så vi kan bruke den igjen nå: 

Denne virker ikke helt kommer ikke til å finne noen sammenhengende, men det må være mulig å fikse. Jeg legger på en atDay her :) 
first.periode.tildato.plusMonths(1) == second.periode.fradato

Reply:
Vi byttet til en fold() versjon. Ble jo mye mindre kode :)